src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -804,6 +812,19 @@ abstract class BTypes {
       "scala/Null",
       "scala/Nothing"
     )
+    def unapply(cr:ClassBType) = Some(cr.internalName)
+
+    def apply(internalName: InternalName, cache: mutable.Map[InternalName, ClassBType])(init: (ClassBType) => Either[NoClassBTypeInfo, ClassInfo]) = {
+      val res = new ClassBType(internalName)
+      // synchronized s required to ensure proper initialisation if info.
+      // see comment on def info
+      res.synchronized {
+        cache(internalName) = res
+        res._info = init(res)
+        res.checkInfoConsistency()
+      }
+      res
```

Comment:
This code doesn't ensure that there's a unique `ClassBType` per `InternalName`. For example, if there are two concurrent executions of `classBTypeFromParsedClassfile` for the same `InternalName`, both could get `None` from `cachedClassBType` and create a new instance.

Reply:
Yes- need to improve that to make the new implementation more tolerent to MT issues than the existing as part of the PR

simplest fix is to use concurrent APIs for the cache, but this still exposes concurrent creation between bTypesFromSymbols and bTypesFromClassFiles

I can't see a logical reason for the separation of the these 2 caches - they perform the same function and are unified for read access. 

@lrytz  Do you have any objection to unifying these caches? Leaving a boolean field (or child class) to signify the origin to satisfy the assert in ` def classBTypeFromSymbol`

At the same time we could remove the scala wrapping of these caches to save a few cpu cycles

apply method would use putIfAbsent to ensure uniqueness. (Ideally we could use computeIfAbsent and remove the whole dummy synchronization, but I think that this would fail due to recursive access)